### ISSUE TITLE
audit our code base where catch is used and see whether list of exceptions are reasonable

### ISSUE THREAD

**Issue by heejaechang** [dotnet/roslyn#297]
some people believe we should either catch all exceptions or not catch at all. so I looked through our code base quickly and it looks like there are several places where we explicitly list some set of exceptions and let others go through.

most common case seems whenever we interact with IO. but there are some other cases where it even catches OOM. 

someone who knows about these better than me should look through those cases and decide whether we should remove those explicit set of exceptions or not.

**Comment by tmat** [comment]
Could you point to the places where the exceptions are listed? 

We have helpers for various IO operations like opening a stream, getting full path, etc. These helpers translate all exceptions that the corresponding BCL APIs throw to IOException, so that the layers above only need to handle a couple of exception types - usually just OperationCanceledException, IOException, BadImageFormatException (if reading metadata). 

We should certainly not be catching OOM.

**Comment by mikedn** [comment]
An OOM is caught here: http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Binder/Binder_Operators.cs,1734

I don't know what kind of code can contain a string literal that's so large to trigger an OOM but it makes sense to translate the OOM to a compiler error in this case. Though I'd put only string.Concat in try/catch, if the OOM comes from allocating a small object like ConstantValue the compiler is toast anyway.

**Comment by tmat** [comment]
I see. I think in this special case OOM is ok to catch - if we try to concat two very long strings and the CLR fails to allocate the result it doesn't mean the process is out of memory. It just means that the string was too long. It's unfortunate that OOM is thrown by String.Concat in that case. Something like ArgumentException would be more appropriate imo.

**Comment by jaredpar** [comment]
The issue that needs to be decided is what is the contract of the method that is performing the operation that results in an exception.  In the cases that have been discussed it has been about methods which have an explicit failure mode in the case an exception performing the operation.  For instance the method would just return `null`, raise an event, etc ...

``` csharp
IFieldSymbol FindFieldSymbol() 
{
    try
    {
       var x = LoadAssemblyInfo();
       ...
       return GetField(x);
     }
     catch (IOException) { return null; } 
     catch (SecurityException) { return null; }
}
```

For these types of methods what is their contract?  Or more specifically what are their possible completion modes?  

1. Succeeded and hence returns a valid value
2. Failed and hence returns an error value (`null`, event raised, etc ...)
3. Throws an exception that indicates ... what precisely?  We already have failed. 

Knowing the possible completion modes is critical for callers to be able to reliably use the function in question.  

In these examples the methods are setup to have all three failure modes because they only handle a subset of the possible exceptions that can be thrown.  The full set which can be thrown from the body of the `try` is unknown.

Sure it is speculated to be only IO related exceptions but that is only speculation.  Knowing the full set would require a very deep dive into the code, often having to trace through virtuals and delegate calls.  Even if it can be proven to be a specific set at a given moment it is virtually guaranteed this set will expand over time.

Taking a step back though such an investigation doesn't seem worthwhile.  Consider what would happen if a deep dive was performed and it turned out that this code also threw `KeyNotFoundException`.  The response would be for the developer to add `KeyNotFoundException` to the set of caught exceptions.  This is true for virtually all exception types because the code isn't concerned with why the operation failed only that it did fail.  

This is why catching `Exception` is a better approach here.  It allows the code to definitively answer the failed / didn't fail aspect of the operation.  Additionally it simplifies callers of this method because they only have to consider two possible completion: succeeded and failed.

**Comment by tmat** [comment]
@jaredpar I agree with everything but the last point. (Almost) never catch all exceptions if the try-block contains a mix of code you wrote and code that you want to guard against failures. If you do you will be hiding your bugs in that code. And it will be hard to find them. No error will get reported, things will just not work. 

Using the example above, suppose "..." contained a read of an array element (that may fail due to index out of range) or a call to a method that calculates something (that may fail due to various reasons) and a catch-all was used. Then even if the call to LoadAssemblyInfo succeeded, but you had a bug in the array index calculation or in your method you'd never know the difference. Because you catch all exceptions regardless where they came from. 

A better way how to write the code above is following:

```csharp
IFieldSymbol FindFieldSymbol() 
{
    AssemblyInfo x;
    try
    {
       x = LoadAssemblyInfo();
    }
    catch
    {
       return null;
    }
    ...
    return GetField(x);
}

Now "..." and GetField are outside of the code that you expect to fail, so any bugs there will be reported. Exceptions thrown there are gonna be unhandled and crash the process for example.

**Comment by jaredpar** [comment]
@tmat I agree with the point you are making there,  the `try / catch` in such a pattern should be much more limited in scope.

**Comment by davkean** [comment]
This doesn't have anything to do with infrastructure as it's a product issue.

Moving to compiler for lack of a better place, and because Jared's currently assigned to it. I'm guessing this isn't on track for 1.1.

**Comment by davkean** [comment]
Putting this in unknown, still don't believe this is an infrastructure issue - not every issue that affects everyone is an infrastructure bug. :)

If individual feature team's want to pick this up and run with it, go ahead, but at the moment there are so many things in infrastructure that would be scheduled above this.

**Comment by jasonmalinowski** [comment]
I'm going to close this. This isn't infrastructure, and I agree that audits are good, I suspect a bug isn't going to motivate that.
